### PR TITLE
Unsafe

### PR DESCRIPTION
**Pull request by nom4dv3**
TODO: use `onlyAscCompile` in `compile`
logic: if (onlyAscCompile(yaml)) { isLocal = true }

### CODE REVIEW

**Comment by nom4dv3** [comment]
Testing note: need lib:unsafe to joint test

File: src/types/zkgyaml_eth.ts [hunk 1]
```diff
@@ -105,7 +109,7 @@ export class EthereumDataSource extends DataSource {
 
   // signaficant to decide which lib dsp main it should use.
   getSignificantKeys() {
```
Comment (murongg) [inline src/types/zkgyaml_eth.ts:111]: Should give here a correct type.
e.g.
```ts
getSignificantKeys(): [DataSourceKind, 'unsafe'] | [DataSourceKind] {
  return this.unsafe ? [this.kind, 'unsafe'] : [this.kind]
}
```
or
```ts
getSignificantKeys(): [DataSourceKind, 'unsafe' | undefined] {
  return this.unsafe ? [this.kind, 'unsafe'] : [this.kind, undefined]
}
```

Comment (murongg) [reply src/types/zkgyaml_eth.ts:111]: `string[]`: unequal length string array, slightly ambiguous

File: src/dsp/ethereum.unsafe-ethereum/dataprep.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import type { EthereumDataPrep } from '../ethereum/blockprep.js'
+import type { EthereumUnsafeDataPrep } from '../ethereum.unsafe/dataprep.js'
+import { DataPrep } from '../interface.js'
+
+// includes both exec & prove params
+export class ETHUnsafe_ETH_DataPrep extends DataPrep {
```
[the remark this answers is not in the document]
Comment (murongg) [reply src/dsp/ethereum.unsafe-ethereum/dataprep.ts:6]: `ETHUnsafeDataPrepETH` or `ETHUnsafeDataPrep`, the second `ETH` is requirement?

[the remark this answers is not in the document]
Comment (nom4dv3) [reply src/dsp/ethereum.unsafe-ethereum/dataprep.ts:6]: yes of course, see the dsp name, ethunsafe, eth, ethunsafe-eth are 3 diff dsps

File: src/types/zkgyaml_eth.ts (hunk 1 above)
Comment (nom4dv3) [reply src/types/zkgyaml_eth.ts]: added in 867a051

File: src/api/compile.ts [hunk 3]
```diff
@@ -71,6 +72,17 @@ export interface CompileResult {
   stats: any
 }
 
+export function onlyAscCompile(yaml: CLEYaml) {
```
[the remark this answers is not in the document]
Comment (nom4dv3) [reply src/api/compile.ts:75]: solved 920c9b1

File: src/dsp/ethereum.unsafe-ethereum/dataprep.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (nom4dv3) [reply src/dsp/ethereum.unsafe-ethereum/dataprep.ts]: solved 23c7a81

**Pull request merged**